### ISSUE TITLE
Restic 0.9.0 failing on Linux/ARM (Backblaze B2 Backend)

### ISSUE THREAD

**Issue by lyderic** [restic/restic#1816]
## Output of `restic version`
```
$ restic version
restic 0.9.0 compiled with go1.10.2 on linux/arm
$ uname -a
Linux balzac 4.3.5-std-1 #1 SMP Fri Feb 19 11:52:18 UTC 2016 armv7l GNU/Linux
$ cat /etc/debian_version
9.4
$ ldd --version
ldd (Debian GLIBC 2.24-11+deb9u3) 2.24
```


## How did you run restic exactly?
```
$ export B2_ACCOUNT_ID=XXX
$ export B2_ACCOUNT_KEY=XXX
$ export RESTIC_PASSWORD=XXX
$ export RESTIC_REPOSITORY=b2:XXX:YYY
$ restic -vvvv snapshots
panic: runtime error: index out of range                                                                                                                                                       
                                                                                                                                                                                               
goroutine 1 [running]:                                                                                                                                                                         
github.com/restic/restic/vendor/github.com/kurin/blazer/x/window.(*Window).sweep(0x10ec0600, 0xa4123f1, 0xbeba3d34, 0xb00c70b8, 0x0, 0x9b47e0)                                                 
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/x/window/window.go:109 +0x568                                                                                              
github.com/restic/restic/vendor/github.com/kurin/blazer/x/window.(*Window).insertAt(0x10ec0600, 0xa4123f1, 0xbeba3d34, 0xb00c70b8, 0x0, 0x9b47e0, 0x56a508, 0x11108ab0)                        
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/x/window/window.go:127 +0xd8                                                                                               
github.com/restic/restic/vendor/github.com/kurin/blazer/x/window.(*Window).Insert(0x10ec0600, 0x56a508, 0x11108ab0)                                                                            
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/x/window/window.go:115 +0x48                                                                                               
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.methodCounter.record(0xf8475800, 0xd, 0x10ec0600, 0x6430f8, 0x12, 0xb1e3179, 0x0, 0xc8)                                             
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/monitor.go:67 +0x80                                                                                                     
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*clientTransport).RoundTrip(0x10eb01a0, 0x10f2ce00, 0x110ab520, 0x10f2ce00, 0x5f56f0)                                              
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/b2.go:178 +0x190                                                                                                        
github.com/restic/restic/vendor/github.com/kurin/blazer/base.makeNetRequest(0x6cc998, 0x110ab520, 0x10f2ce00, 0x6c8a58, 0x10eb01a0, 0x110ee1c8, 0x1, 0x0)                                      
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/base/base.go:290 +0x4c                                                                                                     
github.com/restic/restic/vendor/github.com/kurin/blazer/base.(*b2Options).makeRequest(0x10eb2a80, 0x6cc998, 0x110ab520, 0x6430f8, 0x12, 0x637375, 0x4, 0x10f77680, 0x3a, 0x55c158, ...)        
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/base/base.go:383 +0x350                                                                                                    
github.com/restic/restic/vendor/github.com/kurin/blazer/base.(*Bucket).ListFileNames(0x1108c0f0, 0x6cc998, 0x110ab520, 0x3e8, 0x0, 0x0, 0x110ee080, 0xb, 0x0, 0x0, ...)                        
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/base/base.go:973 +0x214                                                                                                    
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*b2Bucket).listFileNames(0x11104050, 0x6cc998, 0x110ab520, 0x3e8, 0x0, 0x0, 0x110ee080, 0xb, 0x0, 0x0, ...)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/baseline.go:297 +0x68                                         
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*beBucket).listFileNames.func1.1(0x0, 0x0)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:297 +0x84                                        
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.withReauth(0x6cc998, 0x110ab520, 0x6ce698, 0x10ebc120, 0x10ec1380, 0x1112b100, 0x426648)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:690 +0x1c                                             
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*beBucket).listFileNames.func1(0x20c34, 0x10ec1340)                
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:310 +0x178
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.withBackoff(0x6cc998, 0x110ab520, 0x6ce698, 0x10ebc120, 0x10ec1340, 0x8, 0x8)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:671 +0x38
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*beBucket).listFileNames(0x1113e200, 0x6cc998, 0x110ab520, 0x3e8, 0x0, 0x0, 0x110ee080, 0xb, 0x0, 0x0, ...)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:312 +0x140
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*Bucket).ListCurrentObjects(0x110aa280, 0x6cc998, 0x110ab520, 0x3e8, 0x111bb7d4, 0x0, 0x3ab6ad3c, 0x6c3eeb, 0x2b822c, 0x20414, ...)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/b2.go:643 +0x90                                         
github.com/restic/restic/internal/backend/b2.(*b2Backend).List(0x10ec0b00, 0x6cc998, 0x110ab520, 0x636ef9, 0x3, 0x110ab500, 0x0, 0x0)
        src/github.com/restic/restic/internal/backend/b2/b2.go:274 +0x3bc                                   
github.com/restic/restic/internal/backend.(*RetryBackend).List.func1(0x1113faa0, 0x1113faa0)
        src/github.com/restic/restic/internal/backend/backend_retry.go:127 +0xec                     
github.com/restic/restic/vendor/github.com/cenkalti/backoff.RetryNotify(0x111bb950, 0x6ca600, 0x1113faa0, 0x111bb8f8, 0x1113faa0, 0x1113faa0)
        src/github.com/restic/restic/vendor/github.com/cenkalti/backoff/retry.go:37 +0x58
github.com/restic/restic/internal/backend.(*RetryBackend).retry(0x1113fa90, 0x6cc998, 0x110ab460, 0x110ee050, 0x9, 0x111bb950, 0x9, 0x6ca300)
        src/github.com/restic/restic/internal/backend/backend_retry.go:36 +0xf4
github.com/restic/restic/internal/backend.(*RetryBackend).List(0x1113fa90, 0x6cc998, 0x110ab440, 0x636ef9, 0x3, 0x1108ce70, 0x0, 0x0)
        src/github.com/restic/restic/internal/backend/backend_retry.go:126 +0x17c
github.com/restic/restic/internal/repository.SearchKey(0x6cc998, 0x10eb2800, 0x10ec1300, 0x10d1c070, 0x1c, 0x14, 0x1177c, 0x0, 0x0)
        src/github.com/restic/restic/internal/repository/key.go:122 +0x160
github.com/restic/restic/internal/repository.(*Repository).SearchKey(0x10ec1300, 0x6cc998, 0x10eb2800, 0x10d1c070, 0x1c, 0x14, 0x0, 0x0)
        src/github.com/restic/restic/internal/repository/repository.go:515 +0x48
main.OpenRepository(0x10d1c0a2, 0x11, 0x0, 0x0, 0x0, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
        src/github.com/restic/restic/cmd/restic/global.go:352 +0x284
main.runSnapshots(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x10d1c0a2, ...)
        src/github.com/restic/restic/cmd/restic/cmd_snapshots.go:50 +0x30
main.glob..func21(0x9af6b0, 0x10eb61e0, 0x0, 0x1, 0x0, 0x0)
        src/github.com/restic/restic/cmd/restic/cmd_snapshots.go:23 +0x48
github.com/restic/restic/vendor/github.com/spf13/cobra.(*Command).execute(0x9af6b0, 0x10eb61d0, 0x1, 0x1, 0x9af6b0, 0x10eb61d0)
        src/github.com/restic/restic/vendor/github.com/spf13/cobra/command.go:756 +0x360
github.com/restic/restic/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x9ae870, 0x6c3ece, 0x0, 0x20414)
        src/github.com/restic/restic/vendor/github.com/spf13/cobra/command.go:846 +0x20c
github.com/restic/restic/vendor/github.com/spf13/cobra.(*Command).Execute(0x9ae870, 0x23, 0x10e71f5c)
        src/github.com/restic/restic/vendor/github.com/spf13/cobra/command.go:794 +0x1c
main.main()
        src/github.com/restic/restic/cmd/restic/main.go:86 +0x128
```

This is another example, with the **check** command:

```
$ restic check
using temporary cache in /tmp/restic-check-cache-159697945
repository 1a0fd35b opened successfully, password is correct
create exclusive lock for repository
load indexes
panic: runtime error: index out of range

goroutine 9 [running]:
github.com/restic/restic/vendor/github.com/kurin/blazer/x/window.(*Window).sweep(0x10e34700, 0x57d230d8, 0xbeba3fe2, 0x67b32a0f, 0x2, 0x9b47e0)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/x/window/window.go:109 +0x568
github.com/restic/restic/vendor/github.com/kurin/blazer/x/window.(*Window).insertAt(0x10e34700, 0x57d230d8, 0xbeba3fe2, 0x67b32a0f, 0x2, 0x9b47e0, 0x56a508, 0x111320c0)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/x/window/window.go:127 +0xd8
github.com/restic/restic/vendor/github.com/kurin/blazer/x/window.(*Window).Insert(0x10e34700, 0x56a508, 0x111320c0)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/x/window/window.go:115 +0x48
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.methodCounter.record(0xf8475800, 0xd, 0x10e34700, 0x648131, 0x18, 0xc874216, 0x0, 0x1a0)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/monitor.go:67 +0x80
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*clientTransport).RoundTrip(0x10eda1d0, 0x11088180, 0x110c8460, 0x11088180, 0x1caf40)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/b2.go:178 +0x190
github.com/restic/restic/vendor/github.com/kurin/blazer/base.makeNetRequest(0x6cc998, 0x110c8460, 0x11088180, 0x6c8a58, 0x10eda1d0, 0x17, 0x2, 0x0)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/base/base.go:290 +0x4c
github.com/restic/restic/vendor/github.com/kurin/blazer/base.(*Bucket).DownloadFileByName(0x110ac270, 0x6cc998, 0x110c8460, 0x10e2a2d0, 0x4d, 0x989680, 0x0, 0x989680, 0x0, 0x0, ...)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/base/base.go:1093 +0x308
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*b2Bucket).downloadFileByName(0x1112e1a8, 0x6cc998, 0x110c8460, 0x10e2a2d0, 0x4d, 0x989680, 0x0, 0x989680, 0x0, 0x2c, ...)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/baseline.go:333 +0x64
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*beBucket).downloadFileByName.func1.1(0x2, 0x2)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:375 +0x74
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.withReauth(0x6cc998, 0x110c8460, 0x6ce698, 0x10ee2300, 0x11140030, 0xaa, 0x2)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:690 +0x1c
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*beBucket).downloadFileByName.func1(0x20c34, 0x1)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:385 +0x13c
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.withBackoff(0x6cc998, 0x110c8460, 0x6ce698, 0x10ee2300, 0x11140000, 0xffffffff, 0x0)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:671 +0x38
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*beBucket).downloadFileByName(0x1113e3f0, 0x6cc998, 0x110c8460, 0x10e2a2d0, 0x4d, 0x989680, 0x0, 0x989680, 0x0, 0x0, ...)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/backend.go:387 +0x100
github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*Reader).thread.func1(0x10f90120)
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/reader.go:126 +0x164
created by github.com/restic/restic/vendor/github.com/kurin/blazer/b2.(*Reader).thread
        src/github.com/restic/restic/vendor/github.com/kurin/blazer/b2/reader.go:100 +0x2c
```

I also ran:
```
$ restic backup ~
```
---> never completed


## What backend/server/service did you use to store the repository?
Backblaze B2


## Expected behavior
expected snapshot listing or check or backup to complete...


## Actual behavior
I got the above stacktrace for **snapshots** one time out of 3 or 4. I got the expected behaviour the others times. But I could never get **backup** or **check** to complete. **backup** doesn't spit out a stacktrace, it just hangs after a while. Please also note that when the command fails, the lock is not released and I have to release it manually with the **unlock** command. The **unlock** command also sometimes fails with a similar stacktrace starting with _panic: runtime error: index out of range_.


## Steps to reproduce the behavior
I cannot reliably reproduce the problem with **snapshots**. It doesn't always happen. I have tried to run without the '-vvvv' or even only '-v', it doesn't seem to make a difference.


## Do you have any idea what may have caused this?
The stacktrace points a lot to the github.com/kurin/blazer package. Unfortunately, I don't have access to another backend on this server to compare.


## Do you have an idea how to solve the issue?
I have recompiled the restic binary myself directly on my Debian ARM server (i.e. not cross-compiled from my Ubuntu amd64 desktop). I used the most recent master HEAD from github (i.e. not the v0.9.0 tag):

```
$ git log --pretty=oneline -n 1
1892b314f82c0367d6daca12927cfcd749545d8a Merge pull request #1815 from restic/update-blazer
```

This _seems_ to solve the problem. Therefore, there _might_ be a problem with the Linux/ARM binary that is provided on the 0.9.0 release page. With the binary I compiled myself, I could successfully do a backup, whereas with the binary provided on the release page, it would hang for a very long time until I lost patience and killed it (I waiting for about 1 hour, my backup is about 7GB of data).

Here the binary I compiled myself:
[restic.gz](https://github.com/restic/restic/files/2040950/restic.gz)

## Did restic help you or made you happy in any way?
Oh yes! I hated backups before restic. Now, please note that 0.8.3 works perfectly for me on ARM, so I can simply revert to using the previous version for the moment.

**Comment by kurin** [comment]
This is my fault, but the version in HEAD now should fix it. I'm on my phone atm but when I get home I'll link the build.

**Comment by kurin** [comment]
Oh you're already building with #1815. That's the one that has the fix. Please let me know if it blows up or hangs (I suspect B2 itself is having issues this week and thus the hangs but I'm not positive).

**Comment by lyderic** [comment]
I was so focused on the fact that it might be an ARM issue (as no one with the more widely used amd64 reported it) that I ruled out that it might actually be a B2 problem!

B2 having issues would partly explain the erratic behaviour, i.e. I don't get the crash each time. However so far, I didn't get it even once with the binary I built myself, that is indeed built with #1815. Same with the 0.8.3 binary.

I will run further tests later and report possible problems.

**Comment by fd0** [comment]
Thanks for the report, and please keep us posted how it goes! You can also find pre-compiled binaries of the code in the master branch here: https://beta.restic.net/?sort=time&order=desc

**Comment by lyderic** [comment]
It's actually very cool that binaries for the ongoing master branch are provided for all architectures. Thank you!
I do further testing and come back to you.

**Comment by lyderic** [comment]
With the binary from master from [this page](https://beta.restic.net/?sort=time&order=desc):

```
$ restic version
restic 0.9.0 (v0.9.0-8-g2a976d79) compiled with go1.10.2 on linux/arm
```
I tested the following commands:
 - **snapshots**: ok
 - **check**: ok (gave me _Fatal: repository contains errors_, certainly due to the fact that I had to interrupt backups forcibly yesterday)
 - **rebuild-index**: ok
 - **prune**: ok
 - **backup**: ok

As far as I can see the problem is solved with restic-v0.9.0-8-g2a976d79 (as it most probably was with the v0.9.0-7-g1892b314 I compiled yesterday).
Please let me know if you want me to test further or if I shall close this report.
L.

**Comment by fd0** [comment]
Thanks for the feedback, I think we can close this issue for now. It was probably an issue with B2, maybe in combination with the fix for the blazer library (thanks @kurin!).

If it happens again, we can reopen this issue here.
